src/hotspot/share/opto/memnode.cpp

```diff
@@ -2770,9 +2770,461 @@ Node *StoreNode::Ideal(PhaseGVN *phase, bool can_reshape) {
+        // Constant long offset
+        constant_offset += offset->get_long();
+      } else if(int_offset == nullptr && parse_int_offset(offset, int_offset, int_offset_shift)) {
+        // LShiftL( ConvI2L(int_offset), int_offset_shift)
+        int_offset = int_offset->uncast();
+        if (int_offset->Opcode() == Op_AddI && int_offset->in(2)->Opcode() == Op_ConI) {
+          // LShiftL( ConvI2L(int_offset + int_con), int_offset_shift)
+          constant_offset += ((jlong)int_offset->in(2)->get_int()) << int_offset_shift;
+          int_offset = int_offset->in(1);
+        }
+      } else {
+        // All others
+        other_offsets->append(offset);
+      }
+    }
+
+    return ArrayPointer(true, pointer, base, constant_offset, int_offset, int_offset_shift, other_offsets);
+  }
+
+  bool is_adjacent_to(const ArrayPointer& other, const jlong data_size) const {
+    if (!_is_valid || !other._is_valid) { return false; }
+
+    // Offset adjacent?
+    if (this->_constant_offset + data_size != other._constant_offset) { return false; }
+
+    // All other components identical?
+    if (this->_base != other._base ||
+        this->_int_offset != other._int_offset ||
+        this->_int_offset_shift != other._int_offset_shift ||
+        this->_other_offsets->length() != other._other_offsets->length()) {
+      return false;
+    }
+
+    for (int i = 0; i < this->_other_offsets->length(); i++) {
+      Node* o1 = this->_other_offsets->at(i);
+      Node* o2 = other._other_offsets->at(i);
+      if (o1 != o2) { return false; }
+    }
+
+    return true;
+  }
+
+#ifndef PRODUCT
+  void dump() {
+    if (!_is_valid) {
+      tty->print("ArrayPointer[%d %s, invalid]", _pointer->_idx, _pointer->Name());
+      return;
+    }
+    tty->print("ArrayPointer[%d %s, base[%d %s] + %lld",
+               _pointer->_idx, _pointer->Name(),
+               _base->_idx, _base->Name(),
+               (long long)_constant_offset);
+    if (_int_offset != 0) {
+      tty->print(" + I2L[%d %s] << %d",
+                 _int_offset->_idx, _int_offset->Name(), _int_offset_shift);
+    }
+    for (int i = 0; i < _other_offsets->length(); i++) {
+      Node* n = _other_offsets->at(i);
+      tty->print(" + [%d %s]", n->_idx, n->Name());
+    }
+    tty->print_cr("]");
+  }
+#endif
+};
+
+StoreNode* StoreNode::can_merge_primitive_array_store_with_def(PhaseGVN* phase, bool check_use) {
+  int opc = Opcode();
+  assert(opc == Op_StoreB || opc == Op_StoreC || opc == Op_StoreI, "precondition");
+
+  StoreNode* def = in(MemNode::Memory)->isa_Store();
+  if (def == nullptr ||
+      def->Opcode() != opc ||
+      def->adr_type()->isa_aryptr() == nullptr) {
+    return nullptr;
+  }
+
+  StoreNode* s1 = this;
+  StoreNode* s2 = def->as_Store();
+  assert(s1->memory_size() == s2->memory_size(), "same size");
+
+  // Check ctrl compatibility
+  Node* ctrl_s1 = s1->in(MemNode::Control);
+  Node* ctrl_s2 = s2->in(MemNode::Control);
+  if (ctrl_s1 != ctrl_s2) {
```

Comment:
Do you need to check that `ctrl_s1` and `ctrl_s2` are not null? I suppose this could be called on a dying part of the graph during igvn.

Reply:
@rwestrel but then would they not be `TOP` rather than `nullptr`?

Reply:
Maybe. I think the current practice is to be extra careful and assume any input can be null during igvn. What do you think @vnkozlov ?